test/end2end_test.go

```diff
@@ -7416,8 +7419,19 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 					}
 					return
 				}
-				if hframe, ok := frame.(*http2.HeadersFrame); ok {
-					sid = hframe.Header().StreamID
+				sid = 0
+				switch fr := frame.(type) {
+				case *http2.HeadersFrame:
+					if !s.waitForEndStream || fr.StreamEnded() {
+						sid = fr.Header().StreamID
+					}
+
+				case *http2.DataFrame:
+					if s.waitForEndStream && fr.Header().Flags.Has(http2.FlagDataEndStream) {
```

Comment:
Why doesn't this also call `fr.StreamEnded()`?

Reply:
And this if looks like a bug to me at first, then I realized that we don't care about DataFrame if `waitForEndStream==false`.

Add more comments?
Or maybe rename the var to `waitForRPCToFinish`?

Reply:
> Why doesn't this also call fr.StreamEnded()?

Because this is how we do it in our transport and I didn't even check to see if this method exists on data frames.  :)

Added comments.

> maybe rename the var to `waitForRPCToFinish`?

But...the RPC isn't finished until the server does its end.  The server is waiting for the client's end stream.  `waitForEndStream` seems like a better name, but if you have other suggestions LMK.